File: 2024/day3.go
package main

import "regexp"
import "strconv"
import "os"

var mul = regexp.MustCompile(`mul\((\d{1,3}),(\d{1,3})\)`)
var dont = regexp.MustCompile(`(?s)don't\(\).*?(do\(\)|$)`)

func mul_sum(memory string) (sum int) {
	for _, subs := range mul.FindAllStringSubmatch(memory, -1) {
		x, _ := strconv.Atoi(subs[1])
		y, _ := strconv.Atoi(subs[2])
		sum += x * y
	}
	return
}

func day3(path string) (answer1, answer2 int, err error) {
	memory, err := os.ReadFile(path)
	if err != nil {
		return
	}
	answer1 = mul_sum(string(memory))
	answer2 = mul_sum(dont.ReplaceAllString(string(memory), "#"))
	return
}

File: 2024/day3.py
import re

with open("day3.in", "r") as f:
    memory = f.read()

def mul_sum(memory):
    muls = re.findall(r"mul\((\d{1,3}),(\d{1,3})\)", memory)
    return sum(int(mul[0]) * int(mul[1]) for mul in muls)

print(mul_sum(memory))
print(mul_sum(re.sub(r"don't\(\).*?(do\(\)|$)", "#", memory, flags=re.DOTALL)))
